geth/rpc/call_raw.go

```diff
@@ -28,20 +28,31 @@ func (c *Client) CallRaw(body string) string {
 	return c.callRawContext(ctx, json.RawMessage(body))
 }
 
-// jsonrpcMessage represents JSON-RPC request, notification, successful response or
-// error response.
+// jsonrpcMessage represents JSON-RPC message
 type jsonrpcMessage struct {
 	Version string          `json:"jsonrpc"`
-	ID      json.RawMessage `json:"id,omitempty"`
-	Method  string          `json:"method,omitempty"`
-	Params  json.RawMessage `json:"params,omitempty"`
-	Error   *jsonError      `json:"error,omitempty"`
-	Result  json.RawMessage `json:"result,omitempty"`
+	ID      json.RawMessage `json:"id"`
+}
+
+type jsonrpcRequest struct {
+	jsonrpcMessage
+	Method string          `json:"method"`
+	Params json.RawMessage `json:"params,omitempty"`
+}
+
+type jsonrpcSuccessfulResponse struct {
+	jsonrpcMessage
+	Result json.RawMessage `json:"result"`
+}
+
+type jsonrpcErrorResponse struct {
+	jsonrpcMessage
+	Error *jsonError `json:"error"`
```

Comment:
I think we could get rid of using pointer here. In original design, pointer was probably used to make "zero value" for `jsonError` and `omitempty` work nicely, but after splitting response structure into success and error response, it's not needed.

Reply:
Good point. Removed :slightly_smiling_face: